### DISCUSSION TITLE
v2.0.0-rc.5

### DISCUSSION THREAD

**Discussion by decahedron1** [pykeio/ort#266] [Announcements]
## Possibly breaking
- **Pre-built static libraries** (i.e. not `cuda` or `tensorrt`) **are now linked with `/MD` instead of `/MT` on Windows**; i.e. MSVC CRT is no longer statically linked. This should resolve linking issues in some cases (particularly crates using other FFI libraries), but may cause issues for others. I have personally tested this in 2 internal pyke projects that depend on `ort` & many FFI libraries and haven't encountered any issues, but your mileage may vary.

## Definitely breaking
- https://github.com/pykeio/ort/commit/069ddfdd2c1e8af0ea637ac40082ee176ebf870d **`ort` now depends on `ndarray` 0.16.**
- https://github.com/pykeio/ort/commit/e2c454941b944a3bdb0a5df2f11cb997f50c2658 **`wasm32-unknown-unknown` support has been removed.**
  * Getting `wasm32-unknown-unknown` working in the first place was basically a miracle. Hacking ONNX Runtime to work outside of Emscripten took a lot of effort, but recent changes to Emscripten and ONNX Runtime have made this exponentially more difficult. Given I am not adequately versed on ONNX Runtime's internals, the nigh-impossibility of debugging [weird errors](https://github.com/pykeio/ort/issues/260), ~~and the vow I took to write as little C++ as possible ever since I learned Rust~~, it's no longer feasible for me to work on WASM support for `ort`.
  * If you were using `ort` in WASM, I suggest you use and/or support the development of alternative WASM-supporting ONNX inference crates like [`tract`](https://github.com/sonos/tract) or [WONNX](https://github.com/webonnx/wonnx).

## Features
- https://github.com/pykeio/ort/commit/ab293f89a2e5321c5ba4998080fed6a84943e5ad **Update to ONNX Runtime v1.19.0.**
- https://github.com/pykeio/ort/commit/ecf76f9c2bca04bd9b9eead5035906e9b1792a16 **Use the URL hash for downloaded model filenames.** Models previously downloaded & cached with `commit_from_url` will be redownloaded.
- https://github.com/pykeio/ort/commit/9d2551470f40d73e9d678de19aa3ff7416ec543f **Add missing configuration keys for some execution providers.**
- https://github.com/pykeio/ort/commit/733b7fa329f527a6c0c5c0e5b67f53b52cb8c0a5 **New callbacks for the simple `Trainer` API**, just like HF's `TrainerCallbacks`! This allows you to write custom logging/LR scheduling callbacks. See the updated [`train-clm-simple`](https://github.com/pykeio/ort/blob/733b7fa329f527a6c0c5c0e5b67f53b52cb8c0a5/examples/training/examples/train-clm-simple.rs#L16-L35) example for usage details.

## Fixes
- https://github.com/pykeio/ort/commit/1692d11b63f1dda9de0d0a3a5d0dcd8d89223a5d Fix OpenVINO EP option bugs.
- https://github.com/pykeio/ort/commit/08aaa0ffabf597eb97514180460de65bd1bf9938 Fix DirectML output tensor extraction.

<hr />

If you have any questions about this release, we're here to help:
- [Discord: `#💬｜ort-discussions`](https://discord.gg/uQtsNu2xMa)
- [GitHub Discussions](https://github.com/pykeio/ort/discussions)

Love `ort`? Consider [supporting us on Open Collective](https://opencollective.com/pyke-osai) 💖

❤️💚💙💛

<hr /><em>This discussion was created from the release <a href='https://github.com/pykeio/ort/releases/tag/v2.0.0-rc.5'>v2.0.0-rc.5</a>.</em>

**Comment by Esword618** [comment]
How to use `v2.0.0-rc.5` in `Cargo.toml`:
I encountered a problem. How to trigger the download of `download-binaries`? Mine seems not to be triggered. The following is mine:

```
[package]
name = "transformers-study"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.86"
image = "0.25.2"
ndarray = { version = "0.16.0", optional = true }
serde = { version = "1.0.206", features = ["derive"] }
serde_json = "1.0.122"
tokenizers = "0.20.0"
ort = { version = "2.0.0-rc.5", default-features = false, features = ["ndarray", "download-binaries"] }
#ort = {git = "https://mirror.ghproxy.com/https://github.com/pykeio/ort", tag = "v2.0.0-rc.5", default-features = false, features = ["ndarray", "download-binaries"]}
#ort = {git = "https://mirror.ghproxy.com/https://github.com/pykeio/ort", tag = "v2.0.0-rc.5", default-features = false, features = ["ndarray", "download-binaries", "fetch-models"]}
tracing-subscriber = "0.3.18"
once_cell = "1.19.0"
raqote = { version = "0.8", default-features = false }
show-image = { version = "0.14", features = [ "image", "raqote" ] }
pix2text = { version = "0.1.0", path = "src/pix2text" }

[features]
default = ["ndarray"]
```

**Reply by Esword618** [reply]
This is my log:

```
The given version [19] is not supported, only version 1 to 10 is supported in this build.

assertion failed: !api.is_null()
thread 'mixtex::tests::test_interface' panicked at C:\Users\jyj34\.cargo\git\checkouts\ort-675c4f2a84c4db27\92541cd\src\lib.rs:213:25:
assertion failed: !api.is_null()
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\std\src\panicking.rs:652
   1: core::panicking::panic_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\core\src\panicking.rs:72
   2: core::panicking::panic
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\core\src\panicking.rs:146
   3: ort::api::closure$0
             at C:\Users\jyj34\.cargo\git\checkouts\ort-675c4f2a84c4db27\92541cd\src\lib.rs:213
   4: std::sync::once_lock::impl$0::get_or_init::closure$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,ort::api::closure_env$0>
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once_lock.rs:269
   5: std::sync::once_lock::impl$0::initialize::closure$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,std::sync::once_lock::impl$0::get_or_init::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,ort::api::closure_env$0>,never$>
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once_lock.rs:476
   6: std::sync::once::impl$2::call_once_force::closure$0<std::sync::once_lock::impl$0::initialize::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,std::sync::once_lock::impl$0::get_or_init::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtAp
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once.rs:208
   7: std::sys::sync::once::queue::Once::call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\std\src\sys\sync\once\queue.rs:183
   8: std::sync::once::Once::call_once_force<std::sync::once_lock::impl$0::initialize::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,std::sync::once_lock::impl$0::get_or_init::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,ort::api::
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once.rs:208
   9: std::sync::once_lock::OnceLock<core::sync::atomic::AtomicPtr<ort_sys::OrtApi> >::initialize<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,std::sync::once_lock::impl$0::get_or_init::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,ort::api::clo
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once_lock.rs:475
  10: std::sync::once_lock::OnceLock<core::sync::atomic::AtomicPtr<ort_sys::OrtApi> >::get_or_try_init<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,std::sync::once_lock::impl$0::get_or_init::closure_env$0<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,ort::api
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once_lock.rs:357
  11: std::sync::once_lock::OnceLock<core::sync::atomic::AtomicPtr<ort_sys::OrtApi> >::get_or_init<core::sync::atomic::AtomicPtr<ort_sys::OrtApi>,ort::api::closure_env$0>
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\std\src\sync\once_lock.rs:269
  12: ort::api
             at C:\Users\jyj34\.cargo\git\checkouts\ort-675c4f2a84c4db27\92541cd\src\lib.rs:169
  13: ort::environment::EnvironmentBuilder::commit
             at C:\Users\jyj34\.cargo\git\checkouts\ort-675c4f2a84c4db27\92541cd\src\lib.rs:261
  14: pix2text::mixtex::MixTexOnnx::build<std::path::PathBuf,std::path::PathBuf,std::path::PathBuf,std::path::PathBuf>
             at .\src\mixtex.rs:41
  15: pix2text::mixtex::tests::test_interface
             at .\src\mixtex.rs:338
  16: pix2text::mixtex::tests::test_interface::closure$0
             at .\src\mixtex.rs:329
  17: core::ops::function::FnOnce::call_once<pix2text::mixtex::tests::test_interface::closure_env$0,tuple$<> >
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23\library\core\src\ops\function.rs:250
  18: core::ops::function::FnOnce::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library\core\src\ops\function.rs:250
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

**Reply by decahedron1** [reply]
Do you have the `ORT_LIB_LOCATION` environment variable set? Or maybe `pkg-config` is installed?

**Reply by Esword618** [reply]
I have set up ORT_LIB_LOCATION, but pkg-config is not. How should pkg-config be set? My computer is a Windows system.

**Reply by decahedron1** [reply]
`pkg-config` shouldn't be set - where is `ORT_LIB_LOCATION` pointing? If you want to use the binaries downloaded by `ort`, then delete the environment variable, since it's overriding the default behavior which uses the downloaded binaries.

**Reply by Esword618** [reply]
ORT_LIB_LOCATION points to the folder of the link library. I set it to the system environment variable, not the user environment variable. After setting, use cargo build to update it again, but there is still a problem and an error will be reported. Do I need to restart the computer? Or is there something wrong with my update operation?

**Reply by decahedron1** [reply]
> ORT_LIB_LOCATION points to the folder of the link library

Where did this binary come from? The binary being provided is version 1.10, when ort expects 1.19, hence the error. If you want to use `ort`'s downloaded binaries, then **delete** the environment variable. After changing environment variables, you need to restart whichever shell/IDE you were running the `cargo build` command in for the update to propagate.

**Reply by Esword618** [reply]
OK, thank you. I'll try it.

**Reply by Esword618** [reply]
Can a function be exposed to initialize and define the ORT_LIB_LOCATION position?

**Reply by Esword618** [reply]
There is still an error. I downloaded the link `https://parcel.pyke.io/v2/delivery/ortrs/packages/msort-binary/1.19.0/ortrs-msort_static-v1.19.0-x86_64-pc-windows-msvc.tgz` and placed it in a path, but an error occurred. Why is this? I'm very curious: 

I did not encounter this problem in 2.0.0.rc.4
```
libort_sys-ae2417700662ce94.rlib(float_conversion.obj) : error LNK2019: unresolved external symbol __std_reverse_trivially_swappable_1, referenced in function "char * __cdecl std::rotate<char *>(char *,char *,char *)" (??$rotate@PEAD@std@@YAPEADPEAD00@Z)
libort_sys-ae2417700662ce94.rlib(charconv_bigint.obj) : error LNK2001: unresolved external symbol __std_reverse_trivially_swappable_1
libort_sys-ae2417700662ce94.rlib(mutex.obj) : error LNK2019: unresolved external symbol _Thrd_yield, referenced in function "public: __cdecl <lambda_0c7c625e92acbcfe85646242ef483888>::operator()(void)const " (??R<lambda_0c7c625e92acbcfe85646242ef483888>@@QEBA@XZ)
C:\Users\jyj34\Desktop\111\target\debug\deps\pix2text-4bce918a05cfa4a2.exe : fatal error LNK1120: 249 unresolved externals
```

**Reply by decahedron1** [reply]
Those `unresolved external symbol` errors are caused by an outdated version of Visual Studio. Make sure you have VS2022 >= 17.10 installed.

**Reply by Esword618** [reply]
But my VS2022 >= 17.10.
